Proposal: Document defining Gutenberg Principles (#30963)
WordPress/gutenberg: docs/explanations/principles.md

> # Gutenberg Principles

Reviewer: I think "Guiding Principles" would be better.

Author: See 1a648ba.

This document has been written so that those contributing to the Gutenberg project can have some understanding of what guides decisions around what is included in the project (or not).

An important thing to keep at top of mind when reviewing these principles is that they are not isolated from each other. There is a cohesiveness and complementary nature between them that results in a synergistic impact and direction.

When applied, sometimes these principles build on each other much like the idea of constructing a house with a foundation, walls and roof. However, other times the application of the principles results in something that is more similar to a tapestry where threads of each principle are embedded together and evident in the finished feature or design of the project.

> ## Aim for delightful, consistent, accessible user interfaces for empowering content creation and expression.

Reviewer: I adhere to "empowering content creation and expression" but I'm not sure the other concepts capture the right concepts. 🤔

Author: What about something like:

> The user experience is a canvas that empowers content creation and expression through intuitive and exploration friendly interfaces

Author: See d6db504.

_“Optimize for the user” (Matias - [Gutenberg or the Ship of Theseus](https://matiasventura.com/post/gutenberg-or-the-ship-of-theseus/))_

Each of the words in this principle are deliberate:

> ### Delightful

Reviewer 2: As much as I love this word, it's just not a good principle because it is so subjective. It has also been confiscated by commercial projects.

Author: Is it primarily the specific word you object to or also the meaning behind the choice of the word (and the intent of the principle)? If just the word itself, do you have any suggestions about what could be used instead?

Reviewer 3: Delightful does sound a bit like marketing-speak. It's not so much the subjectivity as the underlying idea that we're somehow directing the user's feelings about the product that makes me uneasy.

The word "transparent" popped into my head when reading `The user interfaces should “get out of the way” and support the user`. Transparency both in the sense of clarity (it's clear how to use the interface, thus "supporting" the user instead of making things difficult) but also in the sense of invisibility (it gets "out of the way" so users don't even notice it while working).

Author: You both have shared perspectives on the word "delightful" that I didn't even think of. I was thinking of it in the context of aiming for user experiences that elicit "surprised joy" (my interpretation of "delight"). I also don't think we can escape the reality that we indeed _do_ influence users feelings about the product when building it.  This principle is aimed at thinking how what we build contributes to positive feelings.

With that said, I do see how **transparent** could be a good substitute here for the word delightful and moves more towards the objectiveness of a _principle_ as opposed to the subjectivity of a value or goal.

Reviewer 4: +1 for `transparent` :+1:

Reviewer: "Delightful" doesn't capture what I think this section hints at. "Transparent" is also lacking a bit in concrete meaning.

The way I'd describe this is "A canvas for expression" (more on that here: https://github.com/WordPress/gutenberg/issues/18667)

Author: I'll see if I can capture some of what you wrote in that issue Matías - ❤️ it.

Author: See the changes made in d6db504.  I incorporated much of what Matías articulated in #18667 and this forms the basis for the revised content of the principle which is now simply "A canvas for expression" with the supporting points of this canvas:

- Guiding users around interacting with the content
- Uniformity within diversity
- and Accessbility

The wording around accessibility has also been tweaked in f8a44af which includes not only the work around this for _creators_ of content, but also for _consumers_ of content (to address what was [brought up here](https://github.com/WordPress/gutenberg/pull/30963#discussion_r620308866)).

This also handles the concerns around the word "delightful".

The user interfaces should “get out of the way” and support the user in creating their content. They should lead to experiences where the user (in the vernacular of Marie Kondo) _finds delight_ with each new thing they discover they can do or accomplish.

> ### Consistent

Reviewer: This is not a principle I'd be comfortable stating as a primary guiding force because it's a bit more nuanced. The way I tend to describe this is more along the lines of "Providing uniformity within diversity, as learning the ingredients of the interface happens once but scales to hundreds of blocks."

This is an important design tension.

Author: I like the concept of "providing uniformity within diversity" but I think there's still the importance here of capturing the idea of not just the _block interfaces_ but also the interfaces within the editor? Would you say this statement is a principle that also applies to the editor interfaces?

Also, is this something that you believe should be explicitly called out as it's _own_ principle, or should it be included somehow as a part of [the statement commented on here](https://github.com/WordPress/gutenberg/pull/30963#discussion_r620302895)?

Author: See d6db504.

Any interfaces are consistently applied. Users should be able to _rely on muscle memory and learned patterns_ as they become familiar with how it works.

This also applies to learning _where to find_ different elements that the user might interact with in creating their content.

> ### Accessible

Reviewer: This one feels a bit odd as it's more of a general principle of WordPress at wide and not specific to Gutenberg. If anything, it should build upon the two other axis of "canvas for expression" and "uniform yet diverse" to embrace the fact users interacting with the software have distinct characteristics and preferences for how to use said software.

Reviewer: There's also something important missing here which is how the software can help users build better websites that are semantic and more accessible. That is, _respect for the viewers as well as the creators_.

Author: > That is, respect for the viewers as well as the creators.

This is definitely something that isn't capture well in the current iteration that should be included ❤️ 

> This one feels a bit odd as it's more of a general principle of WordPress at wide and not specific to Gutenberg. If anything, it should build upon the two other axis of "canvas for expression" and "uniform yet diverse" to embrace the fact users interacting with the software have distinct characteristics and preferences for how to use said software.

While I agree it's a general principle of WordPress, I also think it's important to call out the distinctive nature of accessibility in the context of Gutenberg. I like how you reflect on describing this as software accommodating users with "distinct characteristics and preferences for how to use said software". An example of this is reflected in the numerous entry points people have to insert a block. Is there some specific concern here around the use of the word "accessible" itself here?

Author: See f8a44af.

### Empowering

With delightful, consistent, and accessible interfaces we are aiming for empowering users to try and do things that they assumed were beyond their capabilities and reach before. We strive for those “Aha” moments of surprise and wonder as folks discover how they can unleash their creativity.

Example: https://twitter.com/colemank83/status/1371846826664591364

## Curated Extensibility

The use of the word _curated_ expressly refers to implementing extensibility in a way that prioritizes those _using Gutenberg interfaces to create their content_ over developers integrating with the system. As a result, every extensibility interface request is filtered through the question of how it impacts creator workflows.

Implementation details are ephemeral. Practically this means that the traditional WP approach of wrapping things in filters and liberal use of actions is avoided in favor of explicit extension interfaces that expose _functionality_ but hide the _implementation details._ Gutenberg has opinionated APIs that favor stability and functionality for the user over flexibility for the developer.

Extensibility interfaces in Gutenberg favor:

> **Modularity** - the concept of being able to combine blocks or components into different types of layout or content (see “[Embrace the modularity](https://riad.blog/2020/01/28/embrace-the-modularity/)”).

Reviewer: There's some overlap here with https://developer.wordpress.org/block-editor/explanations/architecture/key-concepts/ — it should be linked at least

Author: See 5e744d9.

**Encapsulation** - there are appropriate error handling and boundaries around things that could be broken by things extending the interface.

**Experimentation** - Often new APIs and interfaces clearly start as experimental while reliability and usability of different approaches are being tested. These experiments always begin with determining what problem is solved from the _creator_ perspective and seek to validate how it impacts the user workflow. The intent is to explore what generalizations of the creator's use-case are possible and to discover better abstractions before solidifying around the contract exposed to extension developers.

> ## Progressive Complexity

Reviewer: This needs to be balanced a bit more with a general seek for "simplicity"

Author: Content wise in the description of the principle, or reflected in the principle statement itself?

Author: This one was a bit difficult for me to address. I think in general I understand your feedback to mean we want to avoid the idea of complexity as a _starting point_ for the interactions. I've made a few changes in e7b3960 which I think helps, but it still feels like it could be better expressed.

> _“On the Layers of the Onion”_ (Matias - Gutenberg or the Ship of Theseus)

Reviewer 5: It isn't linked.

Author: Addressed in 5e8d288.

An example here would be creators discovering they can move a block into another block in various contexts and discovering it _just works_. So even learning basic interactions can help _empower_ more complex interactions through the delight of experimenting as a natural outflow of the creative process. A great example of this is the power of multi-block transforms.

There’s also the related application here to design systems in larger organizations where the design team wants to put some restrictions on what can be modified and manipulated by the content creation team in keeping with the chosen designs behind the presentation of that content. In this case, the progressive complexity principle in Gutenberg provides the tools for “locking down” various elements of the editor when and as needed depending on who is working with the content. Entire swaths of the interface could be hidden or removed to account for the fluent needs of those interacting with the content.

> ## Ubiquitous and Safe Adoption

Reviewer: I this is more about a profound respect for the user in an expansive sense — both in how it respects where they come from (compatibility) and aims to ensure protections for the future (the way user data is treated, for example)

Author: So is this more a comment on the principle wording itself (that you would like changed) or on the description of the principle?

Author: I left this as is, it's not clear to me how to improve it.

This encompasses the idea of _backwards compatibility_ and _graceful degradation_.

“One’s existing content should never be lost or generally affected by the switch to a new editor. The newer tool should understand the content natively or at the very least leave it undisturbed.” (Backwards compatibility principle in “[The Language of Gutenberg](https://lamda.blog/2018/04/22/the-language-of-gutenberg/)” - Miguel)

This also encompasses the ideas of _portability and freedom from lock-in_:

> “One’s content shouldn’t be tied to anything. This means that content shouldn’t be tied to any runtime, be it Gutenberg or WordPress in general, and that a good relationship with other editors (mobile apps, MarsEdit, etc.) should be developed.” (Portability principle in “The Language of Gutenberg - Miguel)

Reviewer 5: Link is missing.

Author: Addressed in 5e8d288.

> “Adopting Gutenberg for early testing or curiosity shouldn’t be an irreversible action for one’s content. Disabling Gutenberg shouldn’t result in unreasonably altered content. The same should later apply to third-party blocks. There is a strong desire to fight lock-in, which is not new but is pervasive in a world of proprietary software and software-as-a-service.” (No commitment principle in “The language of Gutenberg - Miguel)

Reviewer 5: Link is missing.

Author: Addressed in 5e8d288.

This also embraces the promise of the Web where content is replicable in multiple ways across devices and apps. It’s the idea that the basic block structure proves to be a timeless and reliable way of representing the content and meta information about that content.

Imagine a world where content created by Gutenberg can be consumed and read _anywhere_. This is largely why at the heart of the block structure, the format is largely an enhancement on top of HTML.

> “Ultimately, choosing HTML means that — as with a painting or a sculpture — the editor’s final artefact is the canonical format of the content, not a byproduct thereof.” (The Language of Gutenberg - Miguel).

Reviewer 5: Link is missing.

Author: Addressed in 5e8d288.

This manifests in letting machines do what machines are good for, and preserve content in a format that is readable to users.

> Finally, the principle of ubiquitous and safe adoption is found in the way Gutenberg is being developed _incrementally_. WordPress powers ~ 40% of the Web and the changes that GB brings cannot happen overnight (Gutenberg or the ship of Theseus - Matías)

Reviewer 5: Link is missing.

And maybe we use "over 40%" so we don't have to change it too often :trollface: 

```suggestion
Finally, the principle of ubiquitous and safe adoption is found in the way Gutenberg is being developed _incrementally_. WordPress powers over 40% of the Web and the changes that GB brings cannot happen overnight (Gutenberg or the ship of Theseus - Matías)
```

### Accessible

Being accessible is a holistic principle that embraces not only the desire that _everyone can work with Gutenberg interfaces no matter their ability_ but also that the improvements made are _equitable_ and improve the experience for both those *creating* the content and those *consuming* the content. This is a lofty challenge and definitely hard balance to get right. As a canvas of expression, the project seeks to embrace the fact that users interacting with the software have distinct characteristics and preferences shaping helpful workflows for what they need to do.

Modern web application development brings its own set of accessibility problems around standards and consistent application between browsers which often requires some unique solutions that might not be apparent on the first attempt. So we acknowledge that accessibility is not something that can be implemented statically but instead is an ongoing part of the interface creation and refinement.

## Curated Extensibility

> The use of the word _curated_ expressly refers to implementing extensibility in a way that prioritizes those _using Gutenberg interfaces to create their content_ over developers integrating with the system. As a result, every extensibility interface request is filtered through the question of how it impacts creator workflows.

Reviewer 6: I'm having a little bit of trouble parsing this paragraph.

In "_using Gutenberg interfaces to create their content_", does `interfaces` refer to UI/user interfaces, or interfaces as a whole, including APIs?

Assuming the prior, maybe "prioritizes those using Gutenberg user interfaces to create their content"?

Author: Good point. Fixed in 9b4eade (which also clarifies the usage of `interfaces` elsewhere in the doc).

Reviewer 6: Looks good; thanks much!

Extensibility interfaces in Gutenberg favor:

**Modularity** - the concept of being able to combine blocks or components into different types of layout or content (see “[Embrace the modularity](https://riad.blog/2020/01/28/embrace-the-modularity/)”).

**Encapsulation** - there are appropriate error handling and boundaries around things that could be broken by things extending the interface.

**Experimentation** - Often new APIs and interfaces clearly start as experimental while reliability and usability of different approaches are being tested. These experiments always begin with determining what problem is solved from the _creator_ perspective and seek to validate how it impacts the user workflow. The intent is to explore what generalizations of the creator's use-case are possible and to discover better abstractions before solidifying around the contract exposed to extension developers.

> For more information around currently built patterns for extensibility and key architectural concepts, see [this document](/docs/explanations/architecture/key-concepts.md)).

Reviewer 5: ```suggestion
For more information around currently built patterns for extensibility and key architectural concepts, see [this document](/docs/explanations/architecture/key-concepts.md).
```

> For more information around currently built patterns for extensibility and key architectural concepts, see [this document](https://developer.wordpress.org/block-editor/explanations/architecture/key-concepts/)).

Reviewer 5: ```suggestion
For more information around currently built patterns for extensibility and key architectural concepts, see [this document](/docs/explanations/architecture/key-concepts.md)).
```

Author: oops, thanks for making these changes to add the relative links @gziolo :)

> ## Progressive Complexity unfolding from simple origins.

Author: Still not happy with this heading. Feels too wordy and abstract.
